gmi_egads/gmi_egads.c

```diff
@@ -0,0 +1,832 @@
+                             int adj_dim, 
+                             int *num_adjacent, 
+                             egads_ent*** adjacent_ents)
+{
+  egads_model *egm = (egads_model*)m;
+
+  int ent_dim = m->ops->dim(m, (struct gmi_ent*)ent);
+  int ent_tag = m->ops->tag(m, (struct gmi_ent*)ent);
+  
+  int pairing = -1;
+  if (adj_dim == 0 && ent_dim == 3)
+    pairing = 0;
+  else if (adj_dim == 1 && ent_dim == 3)
+    pairing = 1;
+  else if (adj_dim == 2 && ent_dim == 3)
+    pairing = 2;
+  else if (adj_dim == 3 && ent_dim == 0)
+    pairing = 3;
+  else if (adj_dim == 3 && ent_dim == 1)
+    pairing = 4;
+  else if (adj_dim == 3 && ent_dim == 2)
+    pairing = 5;
+  else
+    gmi_fail("bad dims in get_3D_adjacency!");
+
+  int *adj_tags = egm->adjacency_table[pairing][ent_tag-1];
+  *num_adjacent = adj_tags[0];
+  *adjacent_ents = (egads_ent**)EG_alloc(sizeof(**adjacent_ents) * (*num_adjacent));
+
+  for (int i = 0; i < *num_adjacent; ++i)
+  {
+    (*adjacent_ents)[i] = (egads_ent*)m->ops->find(m, adj_dim, adj_tags[i+1]);
+  }
+}
+
+/// TODO: implement based on adjacent face's bounding boxes?
+static void get_3D_bounding_box(egads_ent *ent, double *box)
+{
+  (void)ent;
+  (void)box;
+  gmi_fail("3D bounding box not implemented!\n");
+}
+
+/// reparameterize a vertex onto an edge
+static void getVertexT(struct gmi_model* m,
+                       struct gmi_ent* to,
+                       struct gmi_ent* from,
+                       double* t)
+{
+  double diff;
+  double t_range[2];
+  m->ops->range(m, to, 0, &(t_range[0]));
+  double vtx_pnt[3];
+  double p[] = {0, 0};
+  m->ops->eval(m, from, p, &(vtx_pnt[0]));
+  double t_pnt[3];
+  m->ops->eval(m, to, &(t_range[0]), &(t_pnt[0]));
+  diff = sqrt(pow(vtx_pnt[0] - t_pnt[0], 2) + 
+              pow(vtx_pnt[1] - t_pnt[1], 2) + 
+              pow(vtx_pnt[2] - t_pnt[2], 2));
+  if (diff < 0.001)
+  {
+    *t = t_range[0];
+  }
+  else
+  {
+    m->ops->eval(m, to, &(t_range[1]), &(t_pnt[0]));
+    // diff = sqrt(pow(vtx_pnt[0] - t_pnt[0], 2) + 
+    //             pow(vtx_pnt[1] - t_pnt[1], 2) + 
+    //             pow(vtx_pnt[2] - t_pnt[2], 2));
+    *t = t_range[1];
+  }
+  return;
+}
+
+/// TODO: handle degenerate edges (cant eval on them)
+/// function to reparameterize a vertex onto a face
+/// (get the u,v parametric coodinates associated with the vertex)
+static void getVertexUV(struct gmi_model* m,
+                        struct gmi_ent* to,
+                        struct gmi_ent* from,
+                        double to_p[2])
+{
+  struct gmi_set* adj_faces = NULL;
+  struct gmi_set* adj_edges = gmi_adjacent(m, from, 1);
+
+  for (int i = 0; i < adj_edges->n; ++i)
+  {
+    adj_faces = gmi_adjacent(m, adj_edges->e[i], 2);
+    for (int j = 0; j < adj_faces->n; ++j)
+    {
+      if (adj_faces->e[j] == to)
+      {
+        double t;
+        getVertexT(m, adj_edges->e[i], from, &t);
+        m->ops->reparam(m, adj_edges->e[i], &t, to, to_p);
+        goto cleanup;
```

Comment:
Can the code at the `cleanup` marker be moved here?  Is the `goto` necessary?

Reply:
I got rid of the `goto`. I'm not sure why I thought I needed it in the first place.